packages/@aws-cdk/aws-ec2/lib/private/cfn-init-internal.ts

```diff
@@ -0,0 +1,137 @@
+import * as iam from '@aws-cdk/aws-iam';
+import { Construct } from '@aws-cdk/core';
+import { UserData } from '../user-data';
+
+/**
+ * The type of the init element.
+ */
+export enum InitElementType {
+  PACKAGE = 'PACKAGE',
+  GROUP = 'GROUP',
+  USER = 'USER',
+  SOURCE = 'SOURCE',
+  FILE = 'FILE',
+  COMMAND = 'COMMAND',
+  SERVICE = 'SERVICE',
+}
+
+/**
+ * The platform to which the init template applies.
+ */
+export enum InitPlatform {
+  WINDOWS = 'WINDOWS',
+  LINUX = 'LINUX',
+}
+
+/**
+ * Context information passed when an InitElement is being consumed
+ * @internal
+ */
+export interface InitBindOptions {
+  /**
+   * Scope in which to define any resources, if necessary.
+   */
+  readonly scope: Construct;
+
+  /**
+   * Which OS platform (Linux, Windows) the init block will be for.
+   * Impacts which config types are available and how they are created.
+   */
+  readonly platform: InitPlatform;
+
+  /**
+   * Ordered index of current element type. Primarily used to auto-generate
+   * command keys and retain ordering.
+   */
+  readonly index: number;
+
+  /**
+   * Instance role of the consuming instance or fleet
+   */
+  readonly instanceRole: iam.IRole;
+}
+
+/**
+ * A return type for a configured InitElement. Both its CloudFormation representation, and any
+ * additional metadata needed to create the CloudFormation::Init.
+ * @internal
```

Comment:
For posterity, can we add a comment here that this is internal so as to not leak the underlying CF structures?